src/databricks/labs/remorph/reconcile/query_builder/aggregate_query.py

```diff
@@ -0,0 +1,285 @@
+        - query Aggregate rules
+
+        Examples:
+            1.Input :: group_list: [Aggregate(type="Max", agg_cols=["col2", "col3"], group_by_columns=["col1"]),
+                              Aggregate(type="Sum", agg_cols=["col1", "col2"], group_by_columns=["col1"])]
+                           Returns -> SELECT max(col2) AS src_max_col2, max(col3) AS src_max_col3,
+                                                            sum(col1) AS src_sum_col1, sum(col2) AS src_sum_col2
+                                                FROM :tbl
+                                                WHERE col1 IS NOT NULL
+                                                 GROUP BY col1
+        2.
+            group_list: [Aggregate(type="avg", agg_cols=["col4"])]
+            :layer: "tgt"
+            :returns -> SELECT avg(col4) AS tgt_avg_col4 FROM :tbl
+
+        :param group_list: List of Aggregate objects with same Group by columns
+        :return: str - SQL Query
+        """
+        cols_with_mapping: list[exp.Expression] = []
+        # Generates a Single Query for multiple aggregates with the same group_by_columns,
+        #   refer to Example 1
+        query_agg_rules = []
+        processed_rules: dict[str, str] = {}
+        for agg in group_list:
+
+            # Skip duplicate rules
+            # Example: {min_grp1+__+grp2 : col1+__+col2}, key = min_grp1+__+grp2
+            key = f"{agg.type}_{agg.group_by_columns_as_str}"
+            if key in processed_rules:
+                existing_rule = processed_rules.get(key)
+                if existing_rule == agg.agg_columns_as_str:
+                    logger.info(
+                        f"Skipping duplicate rule for key: {key}, value: {agg.agg_columns_as_str},"
+                        f" layer: {self.layer}"
+                    )
+                    continue
+            processed_rules[key] = agg.agg_columns_as_str
+
+            # Get the rules for each aggregate and append to the query_agg_rules list
+            query_agg_rules.extend(self._build_aggregate_rules(agg))
+
+            # Get the mapping with alias for aggregate columns and append to the cols_with_mapping list
+            cols_with_mapping.extend(self._get_mapping_cols_with_alias(agg.agg_columns, agg.type))
+
+        # Apply user transformations on Select columns
+        # Example: {column_name: creation_date, source: creation_date, target: to_date(creation_date,'yyyy-mm-dd')}
+        select_cols_with_transform = (
+            self._apply_user_transformation(cols_with_mapping) if self.user_transformations else cols_with_mapping
+        )
+
+        # Transformed columns
+        select_cols_with_alias = self._agg_query_cols_with_alias(select_cols_with_transform)
+        query_exp = exp.select(*select_cols_with_alias).from_(":tbl").where(self.filter)
+
+        # Apply Group by if group_by_columns are defined
+        if group_list[0].group_by_columns:
```

Comment:
We should check if group_list is empty or not before getting item using index

Reply:
We are grouping on "group_by_columns_as_str" field in the `grouped_aggregates` function, if the group by columns are empty, it's defaulted to "NA".

So here, there is always an item in the list.  Just to be sure, added an assert statement. 

```
@property
    def group_by_columns_as_str(self):
        return "+__+".join(self.group_by_columns) if self.group_by_columns else "NA"
```